File: tests/variable-length-test.py
import test
import sys
import re

import logging
logging.basicConfig(stream=sys.stderr, level=logging.DEBUG)

script = 'challenges/variable-length.py'
CPATH= 'content/textfiles';

def correctOutput(input):
  inputPath= input[0]
  fn= input[1]
  ln= input[2]
  newBday= input[3]
  filecontent= open(inputPath, 'r').read()
  myOutput= re.sub(r'(^|\s+|\W+)'+fn+'\|'+ln+'\|\d+', r"\1"+fn+"|"+ln+"|"+newBday, filecontent);
  return myOutput

  
def validate(input, internalTestOutput, studentTestOutput):
  studentOutput= open(input[0]).read()
  return studentOutput == internalTestOutput
  

# clear the output file to the default values
open('/tmp/v1','w').write(open(CPATH + '/pipe.txt', 'r').read())
open('/tmp/v2','w').write(open(CPATH + '/pipe.txt', 'r').read())
open('/tmp/v3','w').write(open(CPATH + '/poem.txt', 'r').read())

test.test(script, ['/tmp/v1', 'Adam', 'Smith', '11111900'], correctOutput, _validate=validate),
test.test(script, ['/tmp/v2', 'Adam', 'Smithers', '00000000'], correctOutput, _validate=validate),
test.test(script, ['/tmp/v3', 'Adam', 'Smith', '11111111'], correctOutput, _validate=validate);

print ("Well done!")

File: tests/variable-length-test.js
var fs= require('fs');
var test = require('./test-fw.js');

function buildTest(pathi, patho, firstname, lastname, bday){
  var safeOutputPath= '/tmp/' + patho.replace(/\W/g, '')
  
  var startingData= fs.readFileSync(pathi, 'utf8')
  fs.writeFileSync(safeOutputPath, startingData, 'utf8')
  
  var regx= firstname+"[|]"+lastname+"[|]"+"\\d{8}";
  var re= new RegExp(regx);
	var correctOutput= startingData.replace(re, firstname+'|'+lastname+'|'+bday).replace(/\s\s*$/, '')
	return {
		inputs: [safeOutputPath, firstname, lastname, bday],
		outputs: [],
    validate: function(i){
      var actualOutput= fs.readFileSync(i.inputs[0], "utf8").replace(/\s\s*$/, '');
      var resultAsExpected= correctOutput == actualOutput;
      if(!resultAsExpected){
        console.info("ACTUAL\n" + actualOutput)
        console.info("CORRECT\n" + correctOutput)
        console.info("LOCAL: " + actualOutput.localeCompare(correctOutput))
      }
      return resultAsExpected;
		}
	}
}

var script = 'challenges/variable-length.js';

var cpath= 'content/textfiles';

test.tests(script, [
	buildTest(cpath+'/pipe.txt', 'vlength1', 'Adam', 'Smith', '02022222'),
	buildTest(cpath+'/pipe.txt', 'vlength2', 'Adam K', 'Smith', '11111111')
]);
